apigen/src/test/projects/asyncposts/src/test/java/com/disteli/posts/PostsTest.java

```diff
@@ -0,0 +1,212 @@
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import com.google.inject.AbstractModule;
+import com.google.inject.Guice;
+import com.google.inject.Injector;
+import com.google.inject.Key;
+import com.google.inject.TypeLiteral;
+import graphql.ExecutionResult;
+import graphql.GraphQLAsync;
+import graphql.schema.DataFetchingEnvironment;
+import graphql.schema.GraphQLObjectType;
+import graphql.schema.GraphQLSchema;
+import graphql.schema.GraphQLType;
+import org.junit.Before;
+import org.junit.Test;
+
+import static graphql.execution.async.AsyncExecutionStrategy.parallel;
+import static java.util.Collections.emptyMap;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static java.util.stream.Collectors.toList;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+
+public class PostsTest {
+
+  private Injector injector;
+  private GraphQLAsync graphQL;
+  private ObjectMapper om;
+
+  @Before
+  public void setup() throws Exception {
+    injector = setupInjector();
+    initGraphQLAsync(injector, "QueryPosts", "MutatePosts");
+    om = new ObjectMapper();
+    om.enable(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS);
+  }
+
+  @Test
+  public void shouldRetrieveAllPosts() throws Exception {
+    // Given
+    String query = "{posts{title author{firstName lastName}}}";
+
+    // When
+    CompletionStage<ExecutionResult> future =
+      graphQL.executeAsync(query, null, null, emptyMap());
+
+    // Then
+    final String result = printAndExtractValue(future);
+    assertTrue(result.contains("firstName\":\"Brian"));
+    assertTrue(result.contains("firstName\":\"Rahul"));
+  }
+
+  @Test
+  public void shouldCreatePost() throws Exception {
+    // Given
+    String query = "mutation{createPost(post:{title:\"NEW\" authorId:1}){title}}";
+
+    // When
+    CompletionStage<ExecutionResult> future =
+      graphQL.executeAsync(query, null, null, emptyMap());
+
+    // Then
+    final String result = printAndExtractValue(future);
+    assertTrue(result.contains("title\":\"NEW"));
+    assertTrue(result.contains("createPost"));
+  }
+
+  public static class AuthorResolver implements Author.AsyncResolver {
+    private Map<Integer, Author> authors;
+
+    public AuthorResolver(Map<Integer, Author> authors) {
+      this.authors = authors;
+    }
+
+    @Override
+    public CompletableFuture<Object> resolve(final DataFetchingEnvironment env) {
+      if (env.getParentType().getName().equals("Post")) {
+        final Post post = env.getSource();
+        if (post.getAuthor().getClass().equals(Author.Unresolved.class)) {
+          return CompletableFuture.completedFuture(authors.values()
```

Comment:
Why are you iterating over every author to obtain the author with this post id (and your using == instead of equals())?

Seems like this would be simpler, no?

authors.get(post.getAuthor().getId());

...there should also be an NPE check in here.